pkg/storage/kubernetes.go

```diff
@@ -227,3 +239,197 @@ func (t *temporaryError) Temporary() bool {
+				stopM <- struct{}{}
+				return
+			case <-deposed:
+				logging.Debugf("Deposed as leader, shutting down")
+				result <- nil
+				return
+			}
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		ctx, cancel := context.WithCancel(context.Background())
+		res := make(chan error)
+
+		go func() {
+			logging.Debugf("Started leader election")
+			le.Run(ctx)
+			logging.Debugf("Finished leader election")
+			res <- nil
+		}()
+
+		// wait for stop
+		<-stopM
+		// cancel fn(ctx)
+		cancel()
+		result <- (<-res)
+	}()
+
+	wg.Wait()
+	close(stopM)
+	logging.Debugf("IPManagementKubernetes: %v, %v", newip, err)
+	return newip, err
+}
+
+// IPManagementKubernetesUpdate manages k8s updates
+func IPManagementKubernetesUpdate(mode int, ipam *KubernetesIPAM, ipamConf whereaboutstypes.IPAMConfig, containerID string, podRef string) (net.IPNet, error) {
+	logging.Debugf("IPManagement -- mode: %v / containerID: %v / podRef: %v", mode, containerID, podRef)
+
+	now := time.Now()
+
+	var newip net.IPNet
+	// Skip invalid modes
+	switch mode {
+	case whereaboutstypes.Allocate, whereaboutstypes.Deallocate:
+	default:
+		return newip, fmt.Errorf("Got an unknown mode passed to IPManagement: %v", mode)
+	}
+
+	var pool IPPool
+	var err error
+
+	ctx, cancel := context.WithTimeout(context.Background(), RequestTimeout)
+	defer cancel()
+
+	// Check our connectivity first
+	if err := ipam.Status(ctx); err != nil {
+		logging.Errorf("IPAM connectivity error: %v", err)
+		return newip, err
+	}
+
+	// handle the ip add/del until successful
+RETRYLOOP:
+	for j := 0; j < DatastoreRetries; j++ {
+		select {
+		case <-ctx.Done():
+			return newip, nil
+		default:
+			// retry the IPAM loop if the context has not been cancelled
+		}
+
+		pool, err = ipam.GetIPPool(ctx, ipamConf.Range)
+		if err != nil {
+			logging.Errorf("IPAM error reading pool allocations (attempt: %d): %v", j, err)
+			if e, ok := err.(temporary); ok && e.Temporary() {
+				continue
+			}
+			return newip, err
+		}
+
+		reservelist := pool.Allocations()
+		var updatedreservelist []whereaboutstypes.IPReservation
+		switch mode {
+		case whereaboutstypes.Allocate:
+			newip, updatedreservelist, err = allocate.AssignIP(ipamConf, reservelist, containerID, podRef)
+			if err != nil {
+				logging.Errorf("Error assigning IP: %v", err)
+				return newip, err
+			}
+		case whereaboutstypes.Deallocate:
+			updatedreservelist, err = allocate.DeallocateIP(ipamConf.Range, reservelist, containerID)
+			if err != nil {
+				logging.Errorf("Error deallocating IP: %v", err)
+				return newip, err
+			}
+		}
+
+		err = pool.Update(ctx, updatedreservelist)
+		if err != nil {
+			logging.Errorf("IPAM error updating pool (attempt: %d): %v", j, err)
+			if e, ok := err.(temporary); ok && e.Temporary() {
+				continue
+			}
+			break RETRYLOOP
+		}
+		break RETRYLOOP
+	}
+
+	//XXX
+	logging.Debugf("Took: %vms\n", time.Since(now).Milliseconds())
```

Comment:
Please remove this because it is only for debug purpose.

Reply:
Right! And this line causes the openshift img build failure, since it uses golang-1.12 - `time.Milliseconds` is available on golang >= 1.13 as per https://golang.org/pkg/time/#Duration.Milliseconds